### ISSUE TITLE
Enhancement: A Truth Maintenance System like DROOLs

### ISSUE THREAD

**Issue by williamwsmith** [noolsjs/nools#84]
http://blog.athico.com/2010/01/drools-inference-and-truth-maintenance.html?m=1

The concept of the logical assert so you don't have to create the retract() rules for example.  I am sure this is very complex.

http://docs.jboss.org/drools/release/5.2.0.M2/drools-expert-docs/html/ch03.html

**Comment by Kramins** [comment]
I would very much like to see this feature added. +1

**Comment by NikitaIT** [comment]
Yes it can be done, it is actually a trivial function. This works in a similar way to observables and reactivity, I have a suspicion that both drools and nools and rete itself can be viewed as an observable resolution algorithm. This should be studied in more detail.

**Comment by NikitaIT** [comment]
if we look at this in more detail, we can see that logical insert is actually not required for automatic retract. This is just a way to explicitly store the intermediate state of the rule.

**Comment by NikitaIT** [comment]
a much more obvious implementation would be:
```ts
const IsAdult = insertLogical(($p : Person) => $p.age >= 18);
const IsAdultBusPass = insertLogical(IsAdult);
IsAdultBusPass.watch(({ $p }) => doAdultBusPass($p));
```

Instead of:
```ts
rule "Infer Adult"
when
  $p : Person(age >= 18)
then
  insertLogical(new IsAdult($p))
end

rule "Issue Adult Bus Pass"
when
  $p : Person()
    IsAdult($p)
then
  insertLogical(new AdultBusPass($p));
end

rule "Do Adult Bus Pass"
when
  $p : Person()
    AdultBusPass($p)
then
  doAdultBusPass($p));
end
```
